parser-utils/src/value_and_span.rs

```diff
@@ -0,0 +1,282 @@
+            _ => IntoIter::Empty(std::iter::empty()),
+        }
+    }
+
+    fn into_child_stream(self) -> Stream {
+        let span = self.span.clone();
+        Stream(self.into_child_stream_iter(), SpanPosition(span))
+    }
+
+    fn into_atom_stream_iter(self) -> IntoIter {
+        IntoIter::Atom(std::iter::once(self))
+    }
+
+    fn into_atom_stream(self) -> Stream {
+        let span = self.span.clone();
+        Stream(self.into_atom_stream_iter(), SpanPosition(span))
+    }
+}
+
+/// `Of` and `of` allow us to express nested parsers naturally.
+///
+/// For example, `vector().of(many(list()))` parses a vector-of-lists, like [(1 2) (:a :b) ("test") ()].
+///
+/// The "outer" parser `P` and the "nested" parser `N` must be compatible: `P` must produce an
+/// output `edn::ValueAndSpan` which can itself be turned into a stream of child elements; and `N`
+/// must accept the resulting input `Stream`.  This compatibility allows us to lift errors from the
+/// nested parser to the outer parser, which is part of what has made parsing `&'a [edn::Value]`
+/// difficult.
+#[derive(Clone)]
+pub struct Of<P, N>(P, N);
+
+impl<P, N, O> Parser for Of<P, N>
+    where P: Parser<Input=Stream, Output=edn::ValueAndSpan>,
+          N: Parser<Input=Stream, Output=O>,
+{
+    type Input = P::Input;
+    type Output = O;
+    #[inline]
+    fn parse_lazy(&mut self, input: Self::Input) -> ConsumedResult<Self::Output, Self::Input> {
+        use self::FastResult::*;
+
+        match self.0.parse_lazy(input) {
+            ConsumedOk((outer_value, outer_input)) => {
+                match self.1.parse_lazy(outer_value.into_child_stream()) {
+                    ConsumedOk((inner_value, _)) | EmptyOk((inner_value, _)) => ConsumedOk((inner_value, outer_input)),
+                    // TODO: Improve the error output to reference the nested value (or span) in
+                    // some way.  This seems surprisingly difficult to do, so we just surface the
+                    // inner error message right now.  See also the comment below.
+                    EmptyErr(e) | ConsumedErr(e) => ConsumedErr(e),
+                }
+            },
+            EmptyOk((outer_value, outer_input)) => {
+                match self.1.parse_lazy(outer_value.into_child_stream()) {
+                    ConsumedOk((inner_value, _)) | EmptyOk((inner_value, _)) => EmptyOk((inner_value, outer_input)),
+                    // TODO: Improve the error output.  See the comment above.
+                    EmptyErr(e) | ConsumedErr(e) => EmptyErr(e),
+                }
+            },
+            ConsumedErr(e) => ConsumedErr(e),
+            EmptyErr(e) => EmptyErr(e),
+        }
+    }
+
+    fn add_error(&mut self, errors: &mut ParseError<Self::Input>) {
+        self.0.add_error(errors);
+    }
+}
+
+#[inline(always)]
+pub fn of<P, N, O>(p: P, n: N) -> Of<P, Skip<N, Eof<Stream>>>
```

Comment:
Further to our discussion the other day, I think this might best be named `of_only` in order to emphasize the `skip(eof())`. Concur?

```
vector().of_only(Query::variable)
```
?